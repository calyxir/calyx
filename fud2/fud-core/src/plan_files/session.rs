use std::{fs, io};

use camino::{Utf8Path, Utf8PathBuf};

use super::{
    ast::AssignmentList,
    error::ParseError,
    parser::{Lexer, Parser},
};

#[derive(Debug)]
pub struct ParseSession {
    buf: Vec<char>,
    file_path: Option<Utf8PathBuf>,
}

impl ParseSession {
    pub fn from_path(file_path: &Utf8Path) -> io::Result<Self> {
        if !file_path.is_file() {
            let msg = format!("{file_path} is not a file");
            return Err(io::Error::other(msg));
        };

        let buf = fs::read_to_string(file_path)?.chars().collect();
        let owned_path = file_path.to_owned();
        Ok(ParseSession {
            buf,
            file_path: Some(owned_path),
        })
    }

    pub fn with_str_buf(src: &str) -> Self {
        let char_buf = src.chars().collect();
        ParseSession {
            buf: char_buf,
            file_path: None,
        }
    }

    pub fn buf(&self) -> &[char] {
        &self.buf
    }

    pub fn file_path(&self) -> &Option<Utf8PathBuf> {
        &self.file_path
    }

    pub fn parse(&self) -> Result<AssignmentList, ParseError> {
        let mut toks = vec![];
        let mut lexer = Lexer::from_parts(self, 0);
        while lexer.has_next_token() {
            let tok = lexer.next_token()?;
            toks.push(tok);
        }
        let mut parser = Parser::from_parts(toks, self);
        parser.parse()
    }
}
