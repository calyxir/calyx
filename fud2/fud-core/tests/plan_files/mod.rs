use fud_core::plan_files::session::ParseSession;

#[test]
fn basic_parse() {
    let src = "x = input(a, b); z, y = f(x);";
    let sess = ParseSession::with_str_buf(src);
    let ast = sess.parse();
    match ast {
        Ok(ast) => insta::assert_debug_snapshot!(ast),
        Err(e) => panic!("{:?}", e.msg()),
    }
}

#[test]
fn no_args() {
    let src = "x = input();";
    let sess = ParseSession::with_str_buf(src);
    let ast = sess.parse();
    match ast {
        Ok(ast) => insta::assert_debug_snapshot!(ast),
        Err(e) => panic!("{:?}", e.msg()),
    }
}

#[test]
fn single_assignment() {
    let src = "x = input(y);";
    let sess = ParseSession::with_str_buf(src);
    let ast = sess.parse();
    match ast {
        Ok(ast) => insta::assert_debug_snapshot!(ast),
        Err(e) => panic!("{:?}", e.msg()),
    }
}

#[test]
fn unicode_ids() {
    let src = "Æ = Ë(Ð);";
    let sess = ParseSession::with_str_buf(src);
    let ast = sess.parse();
    match ast {
        Ok(ast) => insta::assert_debug_snapshot!(ast),
        Err(e) => panic!("{:?}", e.msg()),
    }
}

#[test]
fn underscores() {
    let src = "__ = _hello__(_Ð);";
    let sess = ParseSession::with_str_buf(src);
    let ast = sess.parse();
    match ast {
        Ok(ast) => insta::assert_debug_snapshot!(ast),
        Err(e) => panic!("{:?}", e.msg()),
    }
}

#[test]
fn many_io_lists() {
    let src = "a, b, c, d, e, f = input(g, h, i, j, k); l, m, n, o = go(p, q, r, s, t); u, v = stop(w, x, y, z);";
    let sess = ParseSession::with_str_buf(src);
    let ast = sess.parse();
    match ast {
        Ok(ast) => insta::assert_debug_snapshot!(ast),
        Err(e) => panic!("{:?}", e.msg()),
    }
}

#[test]
fn weird_whitespace() {
    let src = "\u{2029}a,b,      c = \t\twah\u{2009}(c\n,d\r\n);\n\r\n\t    \te,f=g(h,i);";
    let sess = ParseSession::with_str_buf(src);
    let ast = sess.parse();
    match ast {
        Ok(ast) => insta::assert_debug_snapshot!(ast),
        Err(e) => panic!("{:?}", e.msg()),
    }
}

#[test]
fn ignore_input() {
    let src = "_ = input();";
    let sess = ParseSession::with_str_buf(src);
    let ast = sess.parse();
    match ast {
        Ok(ast) => insta::assert_debug_snapshot!(ast),
        Err(e) => panic!("{:?}", e.msg()),
    }
}

#[test]
fn missing_ident() {
    let src = " = input();";
    let sess = ParseSession::with_str_buf(src);
    let ast = sess.parse();
    match ast {
        Ok(ast) => panic!("this was supposed to be an error, but got: {ast:?}"),
        Err(e) => insta::assert_snapshot!(e.msg()),
    }
}

#[test]
fn empty_file() {
    let src = "";
    let sess = ParseSession::with_str_buf(src);
    let ast = sess.parse();
    match ast {
        Ok(ast) => insta::assert_debug_snapshot!(ast),
        Err(e) => insta::assert_snapshot!(e.msg()),
    }
}

#[test]
fn only_whitespace() {
    let src = "  \n \t ";
    let sess = ParseSession::with_str_buf(src);
    let ast = sess.parse();
    match ast {
        Ok(ast) => insta::assert_debug_snapshot!(ast),
        Err(e) => insta::assert_snapshot!(e.msg()),
    }
}

#[test]
fn missing_value() {
    let src = "x = ;";
    let sess = ParseSession::with_str_buf(src);
    let ast = sess.parse();
    match ast {
        Ok(ast) => insta::assert_debug_snapshot!(ast),
        Err(e) => insta::assert_snapshot!(e.msg()),
    }
}

#[test]
fn missing_assign_operator() {
    let src = "x input();";
    let sess = ParseSession::with_str_buf(src);
    let ast = sess.parse();
    match ast {
        Ok(ast) => insta::assert_debug_snapshot!(ast),
        Err(e) => insta::assert_snapshot!(e.msg()),
    }
}

#[test]
fn missing_arguments() {
    let src = "x = input;";
    let sess = ParseSession::with_str_buf(src);
    let ast = sess.parse();
    match ast {
        Ok(ast) => insta::assert_debug_snapshot!(ast),
        Err(e) => insta::assert_snapshot!(e.msg()),
    }
}

#[test]
fn missing_semicolon() {
    let src = "x = input()";
    let sess = ParseSession::with_str_buf(src);
    let ast = sess.parse();
    match ast {
        Ok(ast) => insta::assert_debug_snapshot!(ast),
        Err(e) => insta::assert_snapshot!(e.msg()),
    }
}

#[test]
fn missing_semicolon_with_more_after() {
    let src = "x = input() y = more();";
    let sess = ParseSession::with_str_buf(src);
    let ast = sess.parse();
    match ast {
        Ok(ast) => insta::assert_debug_snapshot!(ast),
        Err(e) => insta::assert_snapshot!(e.msg()),
    }
}

#[test]
fn trailing_comma_in_vars() {
    let src = "x, = input();";
    let sess = ParseSession::with_str_buf(src);
    let ast = sess.parse();
    match ast {
        Ok(ast) => insta::assert_debug_snapshot!(ast),
        Err(e) => insta::assert_snapshot!(e.msg()),
    }
}

#[test]
fn trailing_comma_in_args() {
    let src = "x, y = input(a, );";
    let sess = ParseSession::with_str_buf(src);
    let ast = sess.parse();
    match ast {
        Ok(ast) => insta::assert_debug_snapshot!(ast),
        Err(e) => insta::assert_snapshot!(e.msg()),
    }
}

#[test]
fn total_gibberish() {
    let src = "x, y, != i39(nput/(a, ),;:";
    let sess = ParseSession::with_str_buf(src);
    let ast = sess.parse();
    match ast {
        Ok(ast) => insta::assert_debug_snapshot!(ast),
        Err(e) => insta::assert_snapshot!(e.msg()),
    }
}

#[test]
fn double_semicolon() {
    let src = "x, y = _XxDarkNightxX_(oops, two, semicolons);;";
    let sess = ParseSession::with_str_buf(src);
    let ast = sess.parse();
    match ast {
        Ok(ast) => insta::assert_debug_snapshot!(ast),
        Err(e) => insta::assert_snapshot!(e.msg()),
    }
}
