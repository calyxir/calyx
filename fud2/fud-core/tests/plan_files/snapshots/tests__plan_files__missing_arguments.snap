---
source: fud2/fud-core/tests/plan_files/mod.rs
expression: e.msg()
---
expected OpenParen but found Semicolon
